File: app/src/main/java/com/robifr/ledger/ui/main/customer/filter/CustomerFilter.java
/**
 * Copyright (c) 2022-present Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.main.customer.filter;

import android.content.DialogInterface;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.robifr.ledger.R;
import com.robifr.ledger.databinding.CustomerDialogFilterBinding;
import com.robifr.ledger.ui.main.customer.CustomerFragment;
import java.util.Objects;

public class CustomerFilter implements DialogInterface.OnDismissListener {
  @NonNull private final CustomerFragment _fragment;
  @NonNull private final CustomerDialogFilterBinding _dialogBinding;
  @NonNull private final BottomSheetDialog _dialog;
  @NonNull private final CustomerFilterBalance _filterBalance;
  @NonNull private final CustomerFilterDebt _filterDebt;

  public CustomerFilter(@NonNull CustomerFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
    this._dialogBinding = CustomerDialogFilterBinding.inflate(this._fragment.getLayoutInflater());

    this._dialog =
        new BottomSheetDialog(this._fragment.requireContext(), R.style.BottomSheetDialog);
    this._dialog.setContentView(this._dialogBinding.getRoot());

    this._filterBalance =
        new CustomerFilterBalance(this._fragment, this._dialogBinding, this._dialog);
    this._filterDebt = new CustomerFilterDebt(this._fragment, this._dialogBinding, this._dialog);

    this._dialog.setOnDismissListener(this);
  }

  @Override
  public void onDismiss(@NonNull DialogInterface dialog) {
    Objects.requireNonNull(dialog);

    this._fragment
        .customerViewModel()
        .onCustomersChanged(this._fragment.customerViewModel().fetchAllCustomers());
    this._fragment
        .customerViewModel()
        .filterView()
        .onFiltersChanged(this._fragment.customerViewModel().filterView().inputtedFilters());

    if (this._dialog.getCurrentFocus() != null) this._dialog.getCurrentFocus().clearFocus();
  }

  @NonNull
  public CustomerFilterBalance filterBalance() {
    return this._filterBalance;
  }

  @NonNull
  public CustomerFilterDebt filterDebt() {
    return this._filterDebt;
  }

  public void openDialog() {
    // Allow bottom sheet to go fully expanded.
    final View bottomSheet =
        Objects.requireNonNull(
            this._dialog.findViewById(com.google.android.material.R.id.design_bottom_sheet));
    bottomSheet.getLayoutParams().height = ViewGroup.LayoutParams.MATCH_PARENT;
    bottomSheet.setLayoutParams(bottomSheet.getLayoutParams());

    this._dialog.getBehavior().setState(BottomSheetBehavior.STATE_HALF_EXPANDED);
    this._dialog.show();
  }
}

File: app/src/main/java/io/github/robifr/ledger/ui/customer/filter/CustomerFilter.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.robifr.ledger.ui.customer.filter

import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import io.github.robifr.ledger.R
import io.github.robifr.ledger.databinding.CustomerDialogFilterBinding
import io.github.robifr.ledger.ui.customer.CustomerFragment

class CustomerFilter(private val _fragment: CustomerFragment) {
  private val _dialogBinding: CustomerDialogFilterBinding =
      CustomerDialogFilterBinding.inflate(_fragment.layoutInflater)
  private val _dialog: BottomSheetDialog =
      BottomSheetDialog(_fragment.requireContext(), R.style.BottomSheetDialog).apply {
        behavior.state = BottomSheetBehavior.STATE_EXPANDED
        setContentView(_dialogBinding.root)
        setOnDismissListener { _fragment.customerViewModel.filterView.onDialogClosed() }
      }
  val filterBalance: CustomerFilterBalance = CustomerFilterBalance(_fragment, _dialogBinding)
  val filterDebt: CustomerFilterDebt = CustomerFilterDebt(_fragment, _dialogBinding)

  init {
    _fragment.fragmentBinding.filtersChip.setText(R.string.customer_filters)
    _fragment.fragmentBinding.filtersChip.setOnClickListener {
      _fragment.customerViewModel.filterView.onDialogShown()
    }
  }

  fun showDialog() {
    _dialog.show()
  }

  fun dismissDialog() {
    _dialog.dismiss()
    _dialog.currentFocus?.clearFocus()
  }
}
